File: app/src/main/java/com/akhbulatov/wordkeeper/presentation/ui/wordcategories/WordCategoriesFragment.java
package com.akhbulatov.wordkeeper.presentation.ui.wordcategories;

import android.app.SearchManager;
import android.content.ComponentName;
import android.content.Context;
import android.os.Bundle;
import android.text.Html;
import android.text.TextUtils;
import android.view.ContextMenu;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;

import com.akhbulatov.wordkeeper.App;
import com.akhbulatov.wordkeeper.R;
import com.akhbulatov.wordkeeper.databinding.FragmentWordCategoriesBinding;
import com.akhbulatov.wordkeeper.domain.global.models.WordCategory;
import com.akhbulatov.wordkeeper.presentation.ui.addeditwordcategory.AddEditWordCategoryDialog;
import com.akhbulatov.wordkeeper.presentation.ui.deletewordcategory.DeleteWordCategoryDialog;
import com.akhbulatov.wordkeeper.presentation.ui.global.base.BaseFragment;
import com.akhbulatov.wordkeeper.presentation.ui.global.list.adapters.WordCategoryAdapter;
import com.akhbulatov.wordkeeper.presentation.ui.global.models.WordCategoryUiModel;
import com.akhbulatov.wordkeeper.presentation.ui.global.models.WordCategoryUiModelKt;
import com.akhbulatov.wordkeeper.presentation.ui.global.views.ContextMenuRecyclerView;
import com.akhbulatov.wordkeeper.ui.activity.MainActivity;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;

public class WordCategoriesFragment extends BaseFragment {

    private WordCategoryAdapter mWordCategoryAdapter;

    @Inject
    ViewModelProvider.Factory viewModelFactory;
    private WordCategoriesViewModel viewModel;

    private FragmentWordCategoriesBinding binding;

    public WordCategoriesFragment() {
        super(R.layout.fragment_word_categories);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        App.appComponent
                .wordCategoriesComponentFactory()
                .create()
                .inject(this);
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        viewModel = new ViewModelProvider(this, viewModelFactory).get(WordCategoriesViewModel.class);
        viewModel.loadWordCategories();

        mWordCategoryAdapter = new WordCategoryAdapter(new WordCategoryAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(@NotNull WordCategory wordCategory) {
                viewModel.onWordCategoryClicked(wordCategory);
            }

            @Override
            public void onMoreOptionsClick(@NotNull View view) {
                view.showContextMenu();
            }
        });
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding = FragmentWordCategoriesBinding.bind(view);
        binding.wordCategoriesRecyclerView.setHasFixedSize(true);
        binding.wordCategoriesRecyclerView.addItemDecoration(new DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL));
        binding.wordCategoriesRecyclerView.setAdapter(mWordCategoryAdapter);
        registerForContextMenu(binding.wordCategoriesRecyclerView);

//        fabAddWord.setOnClickListener(view1 ->
//                mListener.onFabAddWordClick(R.string.add_edit_word_add_title,
//                        R.string.add_edit_word_action_add,
//                        android.R.string.cancel)
//        );

        viewModel.getViewState().observe(getViewLifecycleOwner(), this::renderViewState);
    }

    @Override
    public void onDestroyView() {
        binding.wordCategoriesRecyclerView.setAdapter(null);
        binding = null;
        super.onDestroyView();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.fragment_category, menu);

        MenuItem searchItem = menu.findItem(R.id.menu_search_category);
        SearchView searchView = (SearchView) searchItem.getActionView();

        SearchManager searchManager = (SearchManager) requireContext().getSystemService(Context.SEARCH_SERVICE);
        if (searchManager != null) {
            searchView.setSearchableInfo(
                    searchManager.getSearchableInfo(new ComponentName(requireContext(), MainActivity.class)));
        }
        searchView.setImeOptions(EditorInfo.IME_ACTION_SEARCH);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextChange(String newText) {
//                final Cursor cursor = mCategoryDbAdapter.getAll();
//                final int column = cursor.getColumnIndex(CategoryEntry.COLUMN_NAME);
                if (newText.length() > 0) {
//                    mWordCategoryAdapter.swapCursor(new FilterCursorWrapper(cursor, newText, column));

                    if (mWordCategoryAdapter.getItemCount() == 0) {
                        String escapedNewText = TextUtils.htmlEncode(newText);
                        String formattedNoResults = String.format(
                                getString(R.string.no_results_category), escapedNewText);
                        CharSequence styledNoResults = Html.fromHtml(formattedNoResults);

                        binding.noResultsTextView.setText(styledNoResults);
                        binding.noResultsTextView.setVisibility(View.VISIBLE);
                    } else {
                        binding.noResultsTextView.setVisibility(View.GONE);
                    }
                } else {
//                    mWordCategoryAdapter.swapCursor(cursor);
                    binding.noResultsTextView.setVisibility(View.GONE);
                }
                return true;
            }

            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.menu_add_category) {
            showAddEditWordCategoryDialog(null);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onCreateContextMenu(@NonNull ContextMenu menu, @NonNull View v, ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo);
        requireActivity().getMenuInflater().inflate(R.menu.selected_category, menu);
    }

    @Override
    public boolean onContextItemSelected(@NonNull MenuItem item) {
        ContextMenuRecyclerView.RecyclerContextMenuInfo info =
                (ContextMenuRecyclerView.RecyclerContextMenuInfo) item.getMenuInfo();

        WordCategory wordCategory = mWordCategoryAdapter.getCurrentList().get(info.getPosition());
        switch (item.getItemId()) {
            case R.id.menu_rename_category:
                showAddEditWordCategoryDialog(wordCategory);
                return true;
            case R.id.menu_delete_category:
                showDeleteWordCategoryDialog(wordCategory);
                return true;
            default:
                return super.onContextItemSelected(item);
        }
    }

    private void renderViewState(WordCategoriesViewModel.ViewState viewState) {
        showEmptyProgress(viewState.getEmptyProgress());
        showEmptyData(viewState.getEmptyData());
        showEmptyError(viewState.getEmptyError().getFirst(), viewState.getEmptyError().getSecond());
        showWordCategories(viewState.getWordCategories().getFirst(), viewState.getWordCategories().getSecond());
    }

    private void showEmptyProgress(boolean show) {
//        binding.emptyProgressBar.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    private void showEmptyData(boolean show) {
//        binding.emptyDataTextView.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    private void showEmptyError(boolean show, String message) {
//        binding.errorTextView.setText(message);
//        binding.errorTextView.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    private void showWordCategories(boolean show, List<WordCategory> wordCategories) {
        mWordCategoryAdapter.submitList(wordCategories);
        binding.wordCategoriesRecyclerView.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    private void showAddEditWordCategoryDialog(WordCategory wordCategory) {
        WordCategoryUiModel uiModel = null;
        if (wordCategory != null) {
            uiModel = WordCategoryUiModelKt.toUiModel(wordCategory);
        }
        AddEditWordCategoryDialog.newInstance(uiModel)
                .show(getParentFragmentManager(), null);
    }

    private void showDeleteWordCategoryDialog(WordCategory wordCategory) {
        DeleteWordCategoryDialog.newInstance(WordCategoryUiModelKt.toUiModel(wordCategory))
                .show(getParentFragmentManager(), null);
    }
}

File: app/src/main/java/com/akhbulatov/wordkeeper/presentation/ui/wordcategories/WordCategoriesFragment.kt
package com.akhbulatov.wordkeeper.presentation.ui.wordcategories

import android.app.SearchManager
import android.content.ComponentName
import android.content.Context
import android.os.Bundle
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.widget.SearchView
import androidx.core.text.htmlEncode
import androidx.core.text.parseAsHtml
import androidx.core.view.isVisible
import androidx.fragment.app.setFragmentResultListener
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView
import com.akhbulatov.wordkeeper.R
import com.akhbulatov.wordkeeper.databinding.FragmentWordCategoriesBinding
import com.akhbulatov.wordkeeper.domain.global.models.WordCategory
import com.akhbulatov.wordkeeper.presentation.ui.addeditword.AddEditWordDialog
import com.akhbulatov.wordkeeper.presentation.ui.addeditwordcategory.AddEditWordCategoryDialog
import com.akhbulatov.wordkeeper.presentation.ui.global.base.BaseFragment
import com.akhbulatov.wordkeeper.presentation.ui.global.base.ConfirmDialog
import com.akhbulatov.wordkeeper.presentation.ui.global.list.adapters.WordCategoryAdapter
import com.akhbulatov.wordkeeper.presentation.ui.global.models.toUiModel
import com.akhbulatov.wordkeeper.presentation.ui.global.utils.requireCompatActivity
import com.akhbulatov.wordkeeper.presentation.ui.global.views.ContextMenuRecyclerView.RecyclerContextMenuInfo
import com.akhbulatov.wordkeeper.presentation.ui.main.MainActivity
import javax.inject.Inject

class WordCategoriesFragment : BaseFragment(R.layout.fragment_word_categories) {

    @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
    private val viewModel by viewModels<WordCategoriesViewModel> { viewModelFactory }

    private var _binding: FragmentWordCategoriesBinding? = null
    private val binding get() = _binding!!

    private val wordCategoryAdapter by lazy {
        WordCategoryAdapter(object : WordCategoryAdapter.OnItemClickListener {
            override fun onItemClick(wordCategory: WordCategory) {
                viewModel.onWordCategoryClicked(wordCategory)
            }

            override fun onMoreOptionsClick(view: View) {
                view.showContextMenu()
            }
        })
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        WordCategoriesComponent.create().inject(this)
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        if (savedInstanceState == null) {
            viewModel.loadWordCategories()
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentWordCategoriesBinding.bind(view)
        with(binding) {
            requireCompatActivity().supportActionBar?.setTitle(R.string.word_categories_title)

            wordCategoriesRecyclerView.setHasFixedSize(true)
            wordCategoriesRecyclerView.addItemDecoration(
                DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL)
            )
            wordCategoriesRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    if (dy > 0) {
                        addWordFab.hide()
                    } else {
                        addWordFab.show()
                    }
                }
            })
            wordCategoriesRecyclerView.adapter = wordCategoryAdapter
            registerForContextMenu(wordCategoriesRecyclerView)

            addWordFab.setOnClickListener { showAddWordDialog() }
        }

        viewModel.viewState.observe(viewLifecycleOwner, Observer { renderViewState(it) })
    }

    override fun onDestroyView() {
        binding.wordCategoriesRecyclerView.adapter = null
        _binding = null
        super.onDestroyView()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.word_categories, menu)
        val searchItem = menu.findItem(R.id.menu_search_word_categories)
        val searchView = searchItem.actionView as SearchView
        val searchManager = requireContext().getSystemService(Context.SEARCH_SERVICE) as SearchManager
        searchView.setSearchableInfo(
            searchManager.getSearchableInfo(ComponentName(requireContext(), MainActivity::class.java))
        )
        searchView.imeOptions = EditorInfo.IME_ACTION_SEARCH
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean = false

            override fun onQueryTextChange(newText: String): Boolean {
                viewModel.onSearchWordCategoryChanged(newText)
                return true
            }
        })
        searchItem.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
            override fun onMenuItemActionExpand(item: MenuItem): Boolean = true

            override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                viewModel.onCloseSearchWordCategoryClicked()
                return true
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_add_word_category) {
            showAddEditWordCategoryDialog(null)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
        super.onCreateContextMenu(menu, v, menuInfo)
        requireActivity().menuInflater.inflate(R.menu.selected_word_category, menu)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val info = item.menuInfo as RecyclerContextMenuInfo
        val wordCategory = wordCategoryAdapter.currentList[info.position]
        return when (item.itemId) {
            R.id.menu_rename_word_category -> {
                showAddEditWordCategoryDialog(wordCategory)
                true
            }
            R.id.menu_delete_word_category -> {
                showDeleteWordCategoryDialog(wordCategory)
                true
            }
            else -> super.onContextItemSelected(item)
        }
    }

    private fun renderViewState(viewState: WordCategoriesViewModel.ViewState) {
        showEmptyProgress(viewState.emptyProgress)
        showEmptyData(viewState.emptyData)
        showEmptyError(viewState.emptyError.first, viewState.emptyError.second)
        showEmptySearchResult(viewState.emptySearchResult.first, viewState.emptySearchResult.second)
        showWordCategories(viewState.wordCategories.first, viewState.wordCategories.second)
    }

    private fun showEmptyProgress(show: Boolean) {
        binding.emptyProgressBar.isVisible = show
    }

    private fun showEmptyData(show: Boolean) {
        binding.emptyDataTextView.isVisible = show
    }

    private fun showEmptyError(show: Boolean, message: String?) {
        binding.errorTextView.text = message
        binding.errorTextView.isVisible = show
    }

    private fun showEmptySearchResult(show: Boolean, query: String?) {
        query?.let {
            val emptyResult = String.format(getString(R.string.word_categories_empty_search_result), it.htmlEncode())
            binding.emptySearchResultTextView.text = emptyResult.parseAsHtml()
        }
        binding.emptySearchResultTextView.isVisible = show
    }

    private fun showWordCategories(show: Boolean, wordCategories: List<WordCategory>) {
        wordCategoryAdapter.submitList(wordCategories)
        binding.wordCategoriesRecyclerView.isVisible = show
    }

    private fun showAddEditWordCategoryDialog(wordCategory: WordCategory?) {
        setFragmentResultListener(AddEditWordCategoryDialog.REQUEST_WORD_CATEGORY) { _, bundle ->
            val category = bundle.getString(AddEditWordCategoryDialog.RESULT_WORD_CATEGORY)!!
            if (wordCategory == null) {
                viewModel.onAddWordCategoryClicked(category)
            } else {
                viewModel.onEditWordCategoryClicked(wordCategory.id, category)
            }
        }
        AddEditWordCategoryDialog.newInstance(wordCategory?.toUiModel())
            .show(parentFragmentManager, null)
    }

    private fun showDeleteWordCategoryDialog(wordCategory: WordCategory) {
        setFragmentResultListener(ConfirmDialog.REQUEST_OK) { _, _ ->
            viewModel.onDeleteWordCategoryWithWordsClicked(wordCategory)
        }
        ConfirmDialog.newInstance(
            R.string.delete_word_category_title,
            R.string.delete_word_category_message,
            R.string.msg_action_delete
        )
            .show(parentFragmentManager, null)
    }

    private fun showAddWordDialog() {
        AddEditWordDialog.newInstance()
            .show(parentFragmentManager, null)
    }
}
